command/validate.py

```diff
@@ -1,8 +1,243 @@
+            if value is False:
+                valid = False
+        
+        return valid
+
+    def _crd_validation(self, bundleData):
+        logger.info("Validating custom resource definitions.")
+        valid = True
+
+        crds = bundleData[self.crdKey]
+
+        for crd in crds:
+            if "metadata" in crd:
+                if "name" in crd["metadata"]:
+                    logger.info("Evaluating crd %s", crd["metadata"]["name"])
+                else:
+                    logger.error("crd metadata.name not defined.")
+                    valid = False
+            else:
+                logger.error("crd metadata not defined.")
+                valid = False
+
+            if "apiVersion" not in crd:
+                logger.error("crd apiVersion not defined.")
+                valid = False
+
+            if "spec" not in crd:
+                logger.error("crd spec not defined.")
+                valid = False
+
+        return valid
+
+    def _csv_validation(self, bundleData):
+        valid = True
+        logger.info("Validating cluster service versions.")
+
+        csvs = bundleData[self.csvKey]
+
+        for csv in csvs:
+            if "metadata" in csv:
+                if self._csv_metadata_validation(csv["metadata"]) is False:
+                    valid = False
+            else:
+                logger.error("csv metadata not defined.")
+                valid = False
+
+            if "apiVersion" not in csv:
+                logger.error("csv apiVersion not defined.")
+                valid = False
+
+            if "spec" in csv:
+                if self._csv_spec_validation(csv["spec"], bundleData) is False:
+                    valid = False
+            else:
+                logger.error("csv spec not defined.")
+                valid = False
+
+        return valid
+
+    def _csv_spec_validation(self, spec, bundleData):
+        valid = True
+
+        warnSpecList = ["displayName", "description", "icon", "version", "provider", "maturity"]
+        
+        for item in warnSpecList:
+            if item not in spec:
+                logger.warning("csv spec.%s not defined" % item)
+
+        if "installModes" not in spec:
+            logger.error("csv spec.installModes not defined" % item)
+            valid = False
+
+        if "install" not in spec:
+            logger.error("csv spec.install not defined" % item)
+            valid = False
+
+        if "customresourcedefinitions" in spec:
+            customresourcedefinitions = spec["customresourcedefinitions"]
+
+            crdList = []
+            for crd in bundleData[self.crdKey]:
+                try:
+                    name = crd["metadata"]["name"]
+                    crdList.append(name)
+                except:
+                    pass
+
+            if "owned" not in customresourcedefinitions:
+                logger.error("spec.customresourcedefinitions.owned not defined for csv")
```

Comment:
It's possible for an operator to not use any CRDs.

Reply:
Addressed in latest commit.